### PR TITLE
fix: handling the serialization of special floating point values while serializing any object

### PR DESCRIPTION
**Pull request by desusai7**
# Handling the Serialization of Special Floating Point Values while Serializing any object using Gson

* Serializing Special Floating Point values as explained in the below mapping: 
> {Double.NaN: "NaN", Double.POSITIVE_INFINITY: "Infinity", Double.NEGATIVE_INFINITY: "-Infinity"}
> {Float.NaN: "NaN", Float.POSITIVE_INFINITY: "Infinity", Float.NEGATIVE_INFINITY: "-Infinity"}
* Added new Gson `TypeAdapter`'s for Data types `Double` and `Float` to facilitate the serialization of Special Floating Point Values as defined in the above mappings
* Added new methods `serialize` and `deserialize` to `RudderGson` which will centralize the usage of Gson for serialization and deserialization across the codebase. These methods will safely catch any exceptions thrown while serialization/de-serialization. 
          
## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeManager.java [hunk 1]
```diff
@@ -242,8 +243,10 @@ private void replayMessageQueue() {
             RudderLogger.logDebug(String.format(Locale.US, "RudderDeviceModeManager: replayMessageQueue: replaying old messages with factories. Count: %d", messageIds.size()));
             for (int i = 0; i < messageIds.size(); i++) {
                 try {
-                    RudderMessage message = RudderGson.getInstance().fromJson(messages.get(i), RudderMessage.class);
-                    makeFactoryDump(message, messageIds.get(i), true);
+                    RudderMessage message = RudderGson.deserialize(messages.get(i), RudderMessage.class);
+                    if (message != null) {
+                        makeFactoryDump(message, messageIds.get(i), true);
```
Comment (ChryssaAliferi) [inline core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeManager.java:248]: Maybe this methods needs some renaming, as `dump` to `processMessage` was renamed a bit above. What do you think?

**Review by ChryssaAliferi — DISMISSED** [review]

File: core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeManager.java (hunk 1 above)
Comment (desusai7) [reply core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeManager.java]: Agreed, renamed `dump` as appropriate across the code.

File: core/src/main/java/com/rudderstack/android/sdk/core/RudderServerConfigManager.java [hunk 2]
```diff
@@ -157,8 +160,10 @@ private void sleep(int retryCount) {
         } catch (InterruptedException ex) {
             ReportManager.reportError(ex);
             RudderLogger.logError(String.format(Locale.US, "RudderServerConfigManager: Sleep: Exception while the thread is in sleep %s", ex.getLocalizedMessage()));
+            Thread.currentThread().interrupt();
```
Comment (itsdebs) [inline core/src/main/java/com/rudderstack/android/sdk/core/RudderServerConfigManager.java:163]: Why have we added this?

Comment (desusai7) [reply core/src/main/java/com/rudderstack/android/sdk/core/RudderServerConfigManager.java:163]: this is added to fix the sonarcloud issues. more on it [here](https://rules.sonarsource.com/java/tag/error-handling/RSPEC-2142/)

**Review by ChryssaAliferi — APPROVED** [review]

**Review by itsdebs — APPROVED** [review]

**Pull request merged**